days/day-09/solutions/day09.sml

```diff
@@ -0,0 +1,69 @@
+use "../../../lib/util.sml";
+
+local
+    structure AS = ArraySlice
+in
+
+val N = 25;
+
+fun solve1 xs =
+    let fun hasPair vals z =
+            if AS.length vals <= 1 then
+                false
+            else
+                let val x = AS.sub (vals, 0)
+                    val y = z - x
+                in
+                    if x <> y andalso AS.exists (fn c => y = c) vals then
+                        true
+                    else
+                        hasPair (AS.subslice (vals, 1, NONE)) z
+                end
+        fun go i =
+            let val subArr = AS.slice (xs, i - N, SOME N)
+                val z = Array.sub (xs, i)
+            in
+                if hasPair subArr z then
+                    go (i + 1)
+                else
+                    z
+            end
+    in
+        go N
+    end
+
+val asum = AS.foldl (op +) 0
+fun amax xs = AS.foldl Int.max (AS.sub (xs, 0)) xs
+fun amin xs = AS.foldl Int.min (AS.sub (xs, 0)) xs
+
+(** naive scan, presumably a sliding window solution would be faster **)
```

Comment:
Sliding window protocol ? https://en.wikipedia.org/wiki/Sliding_window_protocol :laughing:

Reply:
I'll send a PR with a faster Python version (in theory) that does this, so that you can see what I mean :)